### PR TITLE
Fetch more references

### PR DESCRIPTION
**Pull request by cdonati**
This addresses the issue of insufficient valid entities being fetched from the datastore.

### CODE REVIEW

File: AppDB/datastore_server.py [hunk 1]
```diff
@@ -2294,19 +2294,57 @@ def __kind_query(self, query, filter_info, order_info):
...
+        current_limit,
+        offset=0,
+        start_inclusive=start_inclusive,
+        end_inclusive=end_inclusive
+      )
+
+      new_entities = self.__fetch_entities(references, clean_app_id(query.app()))
+      entities.extend(new_entities)
+
+      # If we have enough valid entities to satisfy the query, we can stop
```
Comment (obino) [inline AppDB/datastore_server.py]: I think the comment reads better with something like: If we satisfied the original query, we are done.

File: AppDB/datastore_server.py [hunk 2]
```diff
@@ -2294,19 +2294,57 @@ def __kind_query(self, query, filter_info, order_info):
...
+      # fetching more references.
+      if len(entities) >= limit:
+        return entities[:limit]
+
+      # If we received fewer references than we asked for, we do not need to ask
```
Comment (obino) [inline AppDB/datastore_server.py]: I think the rationale here is that if we have fewer entities than references, we had some invalid entities, so we need to fetch more entities. Or something like that?

File: AppDB/datastore_server.py [hunk 3]
```diff
@@ -2294,19 +2294,57 @@ def __kind_query(self, query, filter_info, order_info):
...
+        end_inclusive=end_inclusive
+      )
+
+      new_entities = self.__fetch_entities(references, clean_app_id(query.app()))
+      entities.extend(new_entities)
+
+      # If we have enough valid entities to satisfy the query, we can stop
+      # fetching more references.
+      if len(entities) >= limit:
+        return entities[:limit]
```
Comment (obino) [inline AppDB/datastore_server.py]: This should return something like  entities = [self.default_namespace()] + entities as the other exit point?

File: AppDB/datastore_server.py [hunk 4]
```diff
@@ -1862,28 +1877,65 @@ def __fetch_entities_from_row_list(self, rowkeys, app_id):
...
 
-  def __fetch_entities(self, refs, app_id):
-    """ Given the results from a table scan, get the references.
-    
-    Args: 
-      refs: key/value pairs where the values contain a reference to 
-            the entitiy table.
-      app_id: A string, the application identifier.
+  def __extract_rowkeys_from_refs(self, refs):
+    """ Extract the rowkeys to fetch from a list of references
```
Comment (menivaitsi) [inline AppDB/datastore_server.py]: Add '.' at the end.

File: AppDB/datastore_server.py [hunk 5]
```diff
@@ -3177,71 +3324,70 @@ def composite_v2(self, query, filter_info):
 
     return self.__fetch_entities(index_result, clean_app_id(query.app()))
 
-  def __extract_entities_from_indexes(self, index_result, direction):
-    """ Takes the index values and creates partial entities out of them.
-   
+  def __extract_value_from_index(self, index_entry, direction):
+    """ Takes the index values and returns the value of the property.
```
Comment (menivaitsi) [inline AppDB/datastore_server.py]: Should "index values" be "index value"?

**Pull request merged**